lib/firenest/replicated_state/server.ex

```diff
@@ -0,0 +1,288 @@
+          {:reply, :ok, %{state | handler: handler, remote: remote}}
+      end
+    else
+      {:reply, {:error, :already_present}, state}
+    end
+  end
+
+  def handle_call({:update, key, pid, arg}, _from, state) do
+    %{store: store, handler: handler, remote: remote} = state
+
+    case Store.fetch(store, key, pid) do
+      {:ok, value, delta} ->
+        case Handler.local_update(handler, arg, key, pid, delta, value) do
+          {:put, value, delta, handler} ->
+            remote = Remote.local_update(remote, key, pid, value, delta)
+            store = Store.local_update(store, key, pid, value, delta)
+            {:reply, :ok, %{state | store: store, handler: handler, remote: remote}}
+
+          {:delete, value, delta, handler} ->
+            remote = Remote.local_update(remote, key, pid, value, delta)
+
+            case Store.local_delete(store, key, pid) do
+              # The value returned from update is fresher
+              {:ok, _value, store} ->
+                remote = Remote.local_delete(remote, key, pid)
+                handler = Handler.local_delete(handler, [value])
+                {:reply, :ok, %{state | store: store, handler: handler, remote: remote}}
+
+              {:last_member, _value, store} ->
+                unlink_flush(pid)
+                remote = Remote.local_delete(remote, key, pid)
+                handler = Handler.local_delete(handler, [value])
+                {:reply, :ok, %{state | store: store, handler: handler, remote: remote}}
+            end
+        end
+
+      :error ->
+        {:reply, {:error, :not_present}, state}
+    end
+  end
+
+  def handle_call({:delete, key, pid}, _from, state) do
+    %{store: store, handler: handler, remote: remote} = state
+
+    case Store.local_delete(store, key, pid) do
+      {:ok, value, store} ->
+        remote = Remote.local_delete(remote, key, pid)
+        handler = Handler.local_delete(handler, [value])
+        {:reply, :ok, %{state | store: store, handler: handler, remote: remote}}
+
+      {:last_member, value, store} ->
+        unlink_flush(pid)
+        remote = Remote.local_delete(remote, key, pid)
+        handler = Handler.local_delete(handler, [value])
+        {:reply, :ok, %{state | store: store, handler: handler, remote: remote}}
+
+      :error ->
+        {:reply, {:error, :not_present}, state}
+    end
+  end
+
+  def handle_call({:delete, pid}, _from, state) do
+    %{store: store, handler: handler} = state
+
+    case Store.local_delete(store, pid) do
+      {:ok, deletes, store} ->
+        unlink_flush(pid)
+        handler = Handler.local_delete(handler, deletes)
+        # TODO: how do we handle remote in here?
+        {:reply, :ok, %{state | store: store, handler: handler}}
+
+      :error ->
+        {:reply, {:error, :not_member}, state}
+    end
+  end
+
+  def handle_call({:list, key}, _from, state) do
+    %{store: store} = state
+
+    read = fn -> Store.list(store, key) end
```

Comment:
I think I agree with José that anons being passed around is less preferable since I usually do mod.callback. Is there a reason we capture for a lot of the callbacks?

Reply:
Yes, I think passing MFArgs would at least make it slightly clearer and less error prone.

Reply:
Why is it less error prone? It's guaranteed to be local, since we check this before doing the `GenServer.call` so the issues of distributed problems don't apply and the code is not prepared for hot upgrades anyway, so it won't accommodate for it either way. I also think it's "leaking" less details from the implementation point of view to return a fun than to return a full module.

Reply:
@michalmuskala imagine someone rewrites this to:

     read = fn -> Store.list(state.store, key) end

We are now copying the whole state to the client. I think we should just generally avoid passing functions around between proxies and we should only do so if we really need to.